esp-hal/src/uart/uhci/normal.rs

```diff
@@ -0,0 +1,254 @@
+use core::mem::ManuallyDrop;
+
+use crate::{
+    dma::{asynch::{DmaRxFuture, DmaTxFuture}, Channel, DmaChannelFor, DmaEligible, DmaRxBuffer, DmaTxBuffer}, peripherals, uart::{
+        uhci::{AnyUhci, UhciInternal}, Uart
+    }, Async, Blocking, DriverMode
+};
+
+/// todo
+pub struct Uhci<'d, Dm>
+where
+    Dm: DriverMode,
+{
+    /// todo
+    pub internal: UhciInternal<'d, Dm>,
+}
+
+impl<'d> Uhci<'d, Blocking> {
+    /// todo
+    pub fn new(
+        uart: Uart<'d, Blocking>,
+        uhci: peripherals::UHCI0<'static>,
+        channel: impl DmaChannelFor<AnyUhci<'d>>,
+    ) -> Self {
+        let channel = Channel::new(channel.degrade());
+        channel.runtime_ensure_compatible(&uhci);
+
+        let self_uhci = UhciInternal {
+            uart,
+            uhci: uhci.into(),
+            channel,
+        };
+
+        self_uhci.init();
+
+        Self {
+            internal: self_uhci,
+        }
+    }
+
+    /// todo
+    pub fn into_async(self) -> Uhci<'d, Async> {
+        let internal = self.internal.into_async();
+        Uhci { internal }
+    }
+}
+
+impl<'d> Uhci<'d, Async> {
+    /// todo
+    pub async fn write_dma(
+        mut self,
+        mut tx_buffer: impl DmaTxBuffer,
+    ) -> UhciDmaTxTransfer<'d, Async, impl DmaTxBuffer> {
```

Comment:
> Hmm, I'm unable to create an example using the transfer structs, because they return DmaRxBuffer which has no option to operate on, I'm doing something really wrong, probably

Yes, the issue is here. Don't use `impl DmaTxBuffer`, use an explicitly named generic parameter instead.